src/libraries/Common/src/System/Net/MultiArrayBuffer.cs

```diff
@@ -0,0 +1,319 @@
+
+            _blocks = null;
+            _blockCount = 0;
+            _activeStart = 0;
+            _availableStart = 0;
+        }
+
+        public void Dispose()
+        {
+            _activeStart = 0;
+            _availableStart = 0;
+
+            if (_blocks is not null)
+            {
+                for (int i = 0; i < _blocks.Length; i++)
+                {
+                    if (_blocks[i] is not null)
+                    {
+                        ArrayPool<byte>.Shared.Return(_blocks[i]!);
+                        _blocks[i] = null;
+                    }
+                }
+
+                _blocks = null;
+            }
+        }
+
+        public MultiMemory ActiveMemory => _blocks is null ? MultiMemory.Empty : new MultiMemory(_blocks!, _activeStart, _availableStart - _activeStart);
+
+        public MultiMemory AvailableMemory => _blocks is null ? MultiMemory.Empty : new MultiMemory(_blocks!, _availableStart, _blockCount * BlockSize - _availableStart);
+
+        public void Discard(int byteCount)
+        {
+            Debug.Assert(byteCount >= 0);
+            Debug.Assert(byteCount <= ActiveMemory.Length, $"MultiArrayBuffer.Discard: Expected byteCount={byteCount} <= {ActiveMemory.Length}");
+
+            int oldStartBlock = _activeStart / BlockSize;
+            _activeStart += byteCount;
+            int newStartBlock = _activeStart / BlockSize;
+
+            while (oldStartBlock < newStartBlock)
+            {
+                Debug.Assert(_blocks is not null);
+                Debug.Assert(_blocks[oldStartBlock] is not null, $"Discard: oldStartBlock is null?? byteCount={byteCount}, _activeStart={_activeStart}, oldStartBlock={oldStartBlock}, newStartBlock={newStartBlock}");
+
+                ArrayPool<byte>.Shared.Return(_blocks[oldStartBlock]!);
+                _blocks[oldStartBlock] = null;
+
+                oldStartBlock++;
+            }
+
+            if (_activeStart == _availableStart)
+            {
+                // Small optimization to restart at the beginning of the current block, since we have no active bytes.
+                // Note, we don't try to release buffers in this case. Maybe we should. But if we did,
+                // we'd need handle the fact that there could be more than one here, if we previously grew the buffer enough for that.
+
+                _activeStart = newStartBlock * BlockSize;
+                _availableStart = newStartBlock * BlockSize;
+            }
+        }
+
+        public void Commit(int byteCount)
+        {
+            Debug.Assert(byteCount >= 0);
+            Debug.Assert(byteCount <= AvailableMemory.Length, $"MultiArrayBuffer.Commit: Expected byteCount={byteCount} <= {AvailableMemory.Length}");
+
+            _availableStart += byteCount;
+        }
+
+        // Ensure at least [byteCount] bytes to write to, up to the specified limit
+        public void TryEnsureAvailableSpaceUpToLimit(int byteCount, int limit)
```

Comment:
Every test of this method has limit == int.MaxValue.  What happens if you call it and the required byteCount would actually put you over the limit?  What does the Try prefix mean with no bool returned to indicate whether it was successful?

Reply:
I think I will refactor this a bit, into two methods... one with a limit, one without. If you don't care about a limit then we can avoid a small amount of work here.

And I'll add tests specifically for these methods.

"Try" is a legacy here and I'll drop it.

> What happens if you call it and the required byteCount would actually put you over the limit?

Then we will allocate up to the limit and no more. So if you call this, it's your responsibility to check how much space was made available, since it may be less than you asked for, if you hit the limit.

You could argue that the caller could enforce this themselves, I suppose. This mattered more on ArrayBuffer, since we wanted to limit the size of the actual byte array if you have a limit. But here we are basically just using it to do a Math.Min.